client/src/main/scala/org/http4s/client/Client.scala

```diff
@@ -1,84 +1,156 @@
-    * @return Task which will generate the Response
+import scalaz.stream.Process.eval_
+
+/**
+  * Contains a [[Response]] that needs to be disposed of to free the underlying
+  * HTTP connection.
+  * @param response
+  * @param dispose
+  */
+final case class DisposableResponse(response: Response, dispose: Task[Unit]) {
+  /**
+    * Returns a task to handle the response, safely disposing of the underlying
+    * HTTP connection when the task finishes.
     */
-  def prepare(req: Request): Task[Response]
+  def apply[A](f: Response => Task[A]): Task[A] =
+    f(response).onFinish { case _ => dispose }
+}
 
-  /** Prepare a single request
-    * @param req [[Request]] containing the headers, URI, etc.
-    * @return Task which will generate the Response
+/**
+  * A [[Client]] submits [[Request]]s to a server and processes the [[Response]].
+  *
+  * @param open a service to asynchronously return a [[DisposableResponse]] from
+  *             a [[Request]].  This is a low-level operation intended for client
+  *             implementations and middlewares.
+  *
+  * @param shutdown a Task to shut down this Shutdown this client, closing any
+  *                 open connections and freeing resources
+  */
+final case class Client(open: Service[Request, DisposableResponse], shutdown: Task[Unit]) {
+  /** Submits a request, and provides a callback to process the response.
+    *
+    * @param req The request to submit
+    * @param f A callback for the response to req.  The underlying HTTP connection
+    *          is disposed when the returned task completes.  Attempts to read the
+    *          response body afterward will result in an error.
+    * @return The result of applying f to the response to req
     */
-  final def apply(req: Request): Task[Response] = prepare(req)
+  def fetch[A](req: Request)(f: Response => Task[A]): Task[A] =
+    open.run(req).flatMap(_.apply(f))
+
+  /**
+    * Returns this client as a [[Service]].  All connections created by this
+    * service are disposed on completion of callback task f.
+    *
+    * This method effectively reverses the arguments to [[fetch]], and is
+    * preferred when an HTTP client is composed into a larger Kleisli function,
+    * or when a common response callback is used by many call sites.
+    */
+  def toService[A](f: Response => Task[A]): Service[Request, A] =
+    open.flatMapK(_.apply(f))
+
+  /**
+    * Returns this client as an [[HttpService]].  It is the responsibility of
+    * callers of this service to run the response body to dispose of the
+    * underlying HTTP connection.
+    *
+    * This is intended for use in proxy servers.  [[fetch]], [[fetchAs]],
+    * and [[toService]] are safer alternatives, as their signatures guarantee
+    * disposal of the HTTP connection.
+    */
+  def streaming: HttpService =
+    open.map { case DisposableResponse(response, dispose) =>
+      response.copy(body = response.body ++ eval_(dispose))
```

Comment:
This should use `onComplete` so that it happens regardless of the form of process termination: the current pools wont care but in the case of a capped connection pool we need the `dispose` to be called in order to alert the pool that another 'slot' has opened up.